src/app/features/event/event-form/hashtags/hashtags.service.ts

```diff
@@ -0,0 +1,129 @@
+import { ElementRef, inject, Injectable } from '@angular/core';
+import { FormControl } from '@angular/forms';
+import {
+  MatAutocompleteSelectedEvent,
+  MatAutocompleteTrigger,
+} from '@angular/material/autocomplete';
+import { MatChipInputEvent, MatChipEditedEvent } from '@angular/material/chips';
+import { SnackBarService } from '@shared/services';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class HashtagsService {
+  private snackBarService = inject(SnackBarService);
+  private readonly MAX_CHARS = 50;
+  private readonly MAX_HASHTAG_COUNT = 100;
+  private addedHashtags: string[] = [];
+  private allHashtags: string[] = [
+    'Impreza',
+    'Piwo',
+    'Piłkarzyki',
+    'Pizza',
+    'Pizza 2',
+    'Pizza 3',
+    'Pizza 4',
+    'Pizza 5',
+    'Karaoke',
+    'Bilard',
+    'Urodziny',
+    'BaaaaaaaaaardzoDłuuuuuugiHashtagAwogóletoooooooooo',
+  ];
+
+  get maxCharsValue() {
+    return this.MAX_CHARS;
+  }
+  get addedHashtagsValue() {
+    return this.addedHashtags;
+  }
+  get allHashtagsValue() {
+    return this.allHashtags;
+  }
+
+  add(
+    event: MatChipInputEvent,
+    hashtagCtrl: FormControl,
+    auto: MatAutocompleteTrigger
+  ): void {
+    if (this.addedHashtags.length >= this.MAX_HASHTAG_COUNT) {
+      this.maxHashtagCountInfo();
+    } else {
+      const hashtagStr = (event.value || '').trim().slice(0, this.MAX_CHARS);
+
+      if (hashtagStr.length >= this.MAX_CHARS) {
+        this.maxCharsInfo();
+      }
+      // Add hashtag
+      if (hashtagStr && !this.addedHashtags.includes(hashtagStr) && !auto.activeOption) {
+        this.addedHashtags.push(hashtagStr);
+      } else if (this.addedHashtags.includes(hashtagStr)) {
+        this.hashtagAlreadyAddedInfo(hashtagStr);
+      }
+
+      // Clear the input value
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      event.chipInput!.clear();
+      hashtagCtrl.setValue('');
+    }
+  }
+
+  addFromAutocomplete(
+    event: MatAutocompleteSelectedEvent,
+    hashtagInput: ElementRef<HTMLInputElement>,
+    hashtagCtrl: FormControl
+  ): void {
+    if (this.addedHashtags.length >= this.MAX_HASHTAG_COUNT) {
+      this.maxHashtagCountInfo();
+    } else {
+      const hashtagStr = event.option.viewValue;
+      if (this.addedHashtags.includes(hashtagStr)) {
+        this.hashtagAlreadyAddedInfo(hashtagStr);
+      } else {
+        hashtagInput.nativeElement.value = '';
+        hashtagCtrl.setValue('');
+        this.addedHashtags.push(hashtagStr);
+      }
+    }
```

Comment:
by zmniejszyć ilośc zagnieżdzeń warto zastosować guard clause https://learningactors.com/javascript-guard-clauses-how-you-can-refactor-conditional-logic/

Reply:
zrobiłam refactor kodu, prośba o sprawdzenie ponowna czy nowa wersja jest akceptowalna